### PR TITLE
finished day12

### PR DESCRIPTION
**Pull request by KoenvdBerg**
Hi JP, 

This puzzle I solved using the algorithm of Dijsktra. I previously implemented that algorithm in Scala during the aoc2018. That makes the amount of code for this puzzle small. The code for Dijkstra is implemented in my `utils` scala script, which I import at the top of the file.

What do you think of this?

### CODE REVIEW

File: aoc2022/src/main/scala/day12.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      .fromResource(s"day$day.txt")
+      .getLines
+      .toVector
+
+  def mountainGraph(map: Vector[Char], width: Int): Graph[Int] = 
+    (i: Int) =>
+      val currentHeight: Int = map(i).toInt
+      val neighbours = FlatGrid.neighbours4(i, width, map.length)
+      if neighbours.isEmpty then Map.empty
+      else
```
Comment (jan-pieter) [inline aoc2022/src/main/scala/day12.scala:41]: The `if` statement is redundant here; the implementation below will work fine on an empty list of neighbours I think.

File: aoc2022/src/main/scala/day12.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      .toVector
+
+  def mountainGraph(map: Vector[Char], width: Int): Graph[Int] = 
+    (i: Int) =>
+      val currentHeight: Int = map(i).toInt
+      val neighbours = FlatGrid.neighbours4(i, width, map.length)
+      if neighbours.isEmpty then Map.empty
+      else 
+        val next: Vector[(Int, Int)] = neighbours
+          .filter((n: Int) => math.abs(map(n).toInt) - currentHeight <= 1)
```
Comment (jan-pieter) [inline aoc2022/src/main/scala/day12.scala:43]: > math.abs(map(n).toInt)

Can a `Char` ever be negative? :thinking:

**Review by jan-pieter — APPROVED** [review]
Great that you recognized that you could use Dijkstra to solve this. I've left some minor remarks.

File: aoc2022/src/main/scala/day12.scala (hunk 2 above)
Comment (KoenvdBerg) [reply aoc2022/src/main/scala/day12.scala:43]:
```
scala> -100.toChar
val res12: Char = ﾜ
                                                                                
scala> -90.toChar
val res13: Char = ﾦ
                                                                                
scala> -80.toChar
val res14: Char = ﾰ
```

Seems like negative chars do exists. I don't think these chars ever show up in the English alphabet though.

**Pull request merged**